### PR TITLE
Add NC Tanf

### PR DESCRIPTION
**Pull request by myang1010**
Fixes [#891](https://github.com/PolicyEngine/policyengine-us/issues/891)

### CODE REVIEW

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/main.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+description: North Carolina TANF need standard by number of household members.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/main.yaml:1]:
```suggestion
description: North Carolina account for the following TANF need standard, dependent on the number of household members.
```

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+description: North Carolina TANF need standard monthly allowed income and need difference.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml:1]:
```suggestion
description: North Carolina limits the TANF program to filers with income excess over the need standard under the following amount.
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml:1]: right?

File: policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_countable_earned_income_need.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_countable_earned_income_need.yaml:1]: could we add some more test for edge cases? E.g. en enrolled person etc.

Comment (PavelMakarchuk) [reply policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_countable_earned_income_need.yaml:1]: same for other unit tests

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class nc_tanf(Variable):
+    value_type = float
+    entity = SPMUnit
+    label = "North Carolina TANF"
+    unit = USD
+    definition_period = YEAR
+    defined_for = "nc_tanf_eligible"
+
+    def formula(spm_unit, period, parameters):
+        payment_percentage = 0.5
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf.py:13]: we should make this a policy parameter as well

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]
Thank you for filing! mostly formatting suggestions at this point

**Review by myang1010 — COMMENTED** [review]
Thanks Pavel for the review. I've incorporated the suggestions and corrected one logic error. Also, left two comments below and would like to know your thoughts.

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_child_eligible.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+class nc_tanf_child_eligible(Variable):
+    value_type = bool
+    entity = Person
+    definition_period = YEAR
+    label = "Person-level eligiblity for TANF based on age and school status."
+
+    def formula(person, period, parameters):
+        child_0_17 = person("is_child", period)
+        is_18 = person("age", period) == 18
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_child_eligible.py:12]: We might want to create a general parameter for the 18 years in the `is_person_demographic_tanf_eligible` file but it is ok to replicate here

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                "nc_tanf_countable_gross_unearned_income",
+            ],
+        )
+        need_standard = spm_unit("nc_tanf_need_standard", period)
+
+        household_size = spm_unit("spm_unit_size", period)
+
+        return (
+            need_standard - income
+        ) / MONTHS_IN_YEAR / household_size < monthly_allowed_difference
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py]: I think the easiest way is to separate each of the calculation steps:
monthly_reduced_need_standard = reduced_need_standard / MONTHS_IN_YEAR
....

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from policyengine_us.model_api import *
+
+
+class nc_demographic_tanf_eligible(Variable):
+    value_type = bool
+    entity = SPMUnit
+    definition_period = YEAR
+    label = "NC Demographic eligibility for TANF"
+    documentation = "Whether any person in a family applying for the Temporary Assistance for Needy Families program meets demographic requirements."
+    adds = ["nc_tanf_child_eligible"]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py:10]: remove the adds here

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/income/earned.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+description: North Carolina counts these income sources as earned income under the TANF program.
+
+# need to confirm
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/income/earned.yaml:3]: remove if confirmed

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/income/earned.yaml:3]: same below

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/income/nc_tanf_countable_earned_income_need.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        ).gov.states.nc.ncdhhs.tanf.income.earned_exclusion
+        tanf_enrolled = spm_unit("is_tanf_enrolled", period)
+        annual_flat_exclusion = p.flat * MONTHS_IN_YEAR
+        # needs modify
+        return where(
+            tanf_enrolled,
+            # For enrolled participants, Colorado applies only a percent exclusion.
+            gross_earnings * (1 - p.percent),
+            # For new enrollees, Colorado applies only a flat exclusion.
+            max_(gross_earnings - annual_flat_exclusion, 0),
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/income/nc_tanf_countable_earned_income_need.py:27]:
```suggestion
        # For new enrollees, Colorado applies only a flat exclusion.
        flat_reduced_earnings =  max_(gross_earnings - annual_flat_exclusion, 0)
# For enrolled participants, Colorado applies only a percent exclusion.
        percent_reduced_earnings =   gross_earnings * (1 - p.percent)
        return where(
            tanf_enrolled,
            percent_reduced_earnings,
            flat_reduced_earnings)
```

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                "nc_tanf_countable_gross_unearned_income",
+            ],
+        )
+        need_standard = spm_unit("nc_tanf_need_standard", period)
+        household_size = spm_unit("spm_unit_size", period)
+        reduced_need_standard = max_(need_standard - income, 0)
+
+        return (
+            reduced_need_standard / household_size
+            >= monthly_allowed_difference * MONTHS_IN_YEAR
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py:29]:
```suggestion
        need_standard_fraction = reduced_need_standard / household_size
        difference_threshold = monthly_allowed_difference * MONTHS_IN_YEAR
        return need_standard_fraction >= difference_threshold
```

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    label = "North Carolina TANF need standard"
+    unit = USD
+    definition_period = YEAR
+    defined_for = StateCode.NC
+
+    def formula(spm_unit, period, parameters):
+        # Get number of people in SPM unit.
+        people = spm_unit("spm_unit_size", period)
+        # Cap them at the maximum specified in the tables.
+        capped_people = min_(people, 14).astype(int)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.py:16]: we should parameterize the 14 as well here

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

**Comment by MaxGhenis** [comment]
Could you remove the parameter placeholders? We limit parameters to values in the law. Instead, you can create variables without formulas, e.g. for the earned income exclusion.

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        p = parameters(period).gov.states.nc.ncdhhs.tanf.need_standard
+        # Get number of people in SPM unit.
+        people = spm_unit("spm_unit_size", period)
+        # Get household size list
+        household_size_list = list(map(int, (p.main)))
+        # Get the maximun number of people defined in the tables.
+        max_people_on_file = max(household_size_list)
+        # Cap them at the maximum specified in the tables.
+        capped_people = min_(people, max_people_on_file).astype(int)
+        # Calculate additional people beyond the maximum in tables.
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.py:23]:
```suggestion
        household_size = spm_unit("spm_unit_size", period)
        # Get household size list
        household_size_list = list(map(int, (p.main)))
        # Get the maximun number of people defined in the tables.
        max_standard_household_size = max(household_size_list)
        # Cap them at the maximum specified in the tables.
        capped_household_size = min_(people, max_people_on_file).astype(int)
        # Calculate additional people beyond the maximum in tables.
```

File: policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_need_standard.yaml:1]: could you add a test for size 1 and 14 - just for edge case pruposes

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+class nc_demographic_tanf_eligible(Variable):
+    value_type = bool
+    entity = SPMUnit
+    definition_period = YEAR
+    label = "NC Demographic eligibility for TANF"
+    documentation = "Whether any person in a family applying for the Temporary Assistance for Needy Families program meets demographic requirements."
+
+    def formula(spm_unit, period, parameters):
+        person = spm_unit.members
+        eligible = person("is_child", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py:13]:
```suggestion
        is_child = person("is_child", period)
```

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
+from policyengine_us.model_api import *
+
+
+class nc_demographic_tanf_eligible(Variable):
+    value_type = bool
+    entity = SPMUnit
+    definition_period = YEAR
+    label = "NC Demographic eligibility for TANF"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py:8]:
```suggestion
    label = "North Carolina Demographic eligibility for TANF"
```

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        income = add(
+            spm_unit,
+            period,
+            [
+                "nc_tanf_countable_earned_income",
+                "nc_tanf_countable_gross_unearned_income",
+            ],
+        )
+
+        return max_((need_standard - income) * payment_percentage, 0)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf.py:26]:
```suggestion
        reduced_need_standard = max_(need_standard - income, 0)
        return reduced_need_standard * payment_percentage
```

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/income/nc_tanf_countable_earned_income.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
+from policyengine_us.model_api import *
+
+
+class nc_tanf_countable_earned_income(Variable):
+    value_type = float
+    entity = SPMUnit
+    label = "North Carolina TANF countable earned income"
+    unit = USD
+    definition_period = YEAR
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/income/nc_tanf_countable_earned_income.py:9]:
```suggestion
    definition_period = YEAR
    defined_for = StateCode.NC 
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/nc/ncdhhs/tanf/income/nc_tanf_countable_earned_income.py:9]: can we add this to all vars that are not defined fro an eligibility variable, just to minimize the computation for other states

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml:1]: rename to `average_reduced_need_standard_thresold.yaml`

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+description: North Carolina uses this threshold amount to determine if a filer is qualified for the TANF program.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml:1]:
```suggestion
description: North Carolina limits the TANF program to filers with an average need standard of each household member, reduced by income, of this amount.
```

File: policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+
+values:
+  2024-01-01: 25
+
+metadata:
+  unit: currency-USD
+  period: month
+  label: North Carolina TANF monthly eligible threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/ncdhhs/tanf/need_standard/monthly_allowed_difference.yaml:9]:
```suggestion
  label: North Carolina TANF average need standard threshold
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

**Comment by PavelMakarchuk** [comment]
Should be the last batch of format adjustments, thank you @myang1010

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+            spm_unit,
+            period,
+            [
+                "nc_tanf_countable_earned_income",
+                "nc_tanf_countable_gross_unearned_income",
+            ],
+        )
+        need_standard = spm_unit("nc_tanf_need_standard", period)
+        household_size = spm_unit("spm_unit_size", period)
+        reduced_need_standard = max_(need_standard - income, 0)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_tanf_income_eligible.py:25]: create a separate policyengine variable for this since it's used both here and in nc_tanf

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Some of these suggestions will apply to other parts of the code.

By the way, I got some of these suggestions from Claude, where I uploaded the program document and used https://gh-stitcher.streamlit.app to compare to this PR. [Here's the exchange.](https://gist.github.com/MaxGhenis/741f8509421f0fe56034319ac70f9679)

File: policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class nc_demographic_tanf_eligible(Variable):
+    value_type = bool
+    entity = SPMUnit
+    definition_period = YEAR
+    label = "North Carolina Demographic eligibility for TANF"
+    documentation = "Whether any person in a family applying for the Temporary Assistance for Needy Families program meets demographic requirements."
+    defined_for = StateCode.NC
+
```
[the remark this answers is not in the document]
Comment (myang1010) [reply policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py:11]: Our current definition of `is_child`, "return person("age", period) < 18" meets NC's definition of child. Do I still need to make it policy parameter?

[the remark this answers is not in the document]
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/nc/ncdhhs/tanf/nc_demographic_tanf_eligible.py:11]: Yes - if a jurisdiction defines a policy parameter we add it as its own parameter. This lets us model changes within their purview.

File: policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_integration.yaml
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_integration.yaml:1]: rename integration.yaml

@PavelMakarchuk could you take a look at the cause for these test failures?

Comment (myang1010) [reply policyengine_us/tests/policy/baseline/gov/states/nc/ncdhhs/tanf/nc_tanf_integration.yaml:1]: Okay, rename the file to "integration.yaml".
Thanks @PavelMakarchuk , `make install` helps and unblocked me from testing in my local!

**Review by MaxGhenis — APPROVED** [review]
